Notation.Plugin.AzureKeyVault/Certificate/CertificateChain.cs

```diff
@@ -8,19 +9,102 @@ namespace Notation.Plugin.AzureKeyVault.Certificate
         {
-            X509Certificate2Collection chain = new X509Certificate2Collection(leafCert);
-            chain.AddRange(certificateBundle);
-            return chain.Select(x => x.RawData).ToList();
+            /*
+            Valid case:
+            1. self-signed leaf cert
+            2. cert chain: CA1 -> CA2 -> leaf
+
+            Invalid cases:
+            1. empty
+            2. non-self-signed leaf cert
+            3. missing intermediate cert: CA1 -> (x) -> leaf 
+            4. missing root cert: (x) -> CA2 -> leaf
+            5. unnecessary intermediate cert: CA1 -> CA2 -> leaf
+                                              (x) -> CA3
+            6. unnecessary root: CA1 -> CA2 -> leaf
+                                 CA3
+            6. unnecessary cert: CA1 -> CA2 -> leaf
+                                 (x) -> CA3
+            7. duplicated certs: CA1 -> CA2 -> leaf
+                                  └───> CA2
+            8. multiple leaf certs: CA1 -> CA2 -> leaf1
+                                            └───> leaf2
+            9. multple chain: CA1 -> CA2 -> leaf1
+                              CA3 -> leaf2
+            */
+            if (certs.Count == 0)
+            {
+                throw new PluginException("The certificates that need to be built into a chain are empty.");
+            }
+
+            if (certs.Count == 1)
+            {
+                if (certs[0].SubjectName.Name != certs[0].IssuerName.Name)
+                {
+                    throw new PluginException("Only obtained one certificate but it is not a self-signed certificate. Please complete the certificate bundle by `ca_certs` through plugin config.");
+                }
+                return certs;
+            }
+
+            // subject distinguished name -> certificate map
+            var certMap = new Dictionary<string, X509Certificate2>();
+            var issuerSet = new HashSet<string>();
+            foreach (var cert in certs)
+            {
+                if (certMap.ContainsKey(cert.SubjectName.Name))
+                {
+                    throw new PluginException($"Found duplicated certificates: {cert.SubjectName.Name}");
+                }
+                certMap[cert.SubjectName.Name] = cert;
+                issuerSet.Add(cert.IssuerName.Name);
+            }
+
+            // count the leaf certificate
+            if (certs.Count(x => !issuerSet.Contains(x.SubjectName.Name)) != 1)
+            {
+                // AKV certificates always contain the leaf certificate
+                throw new PluginException("Found multiple leaf certificates or unreferenced intermediate CAs");
+            }
+            var leafCert = certs.First(x => !issuerSet.Contains(x.SubjectName.Name));
+
+            // build the certificate chain
+            X509Certificate2Collection chain = new X509Certificate2Collection();
+            var currentCert = leafCert;
+            while (currentCert != null)
```

Comment:
What if a special cert chain is crafted as follows where `A` is a leaf cert.
```mermaid
graph LR;
C -- issue --> B
B -- issue --> A
B -- issue --> C
```
What will happen? Will it happen? Will it be dead loop?

Reply:
Nice catch! I thought about the cycle before, but I thought it couldn't generate a chain with cycle, I should have thought about it more carefully.

Added 
```go
if (!caSet.Remove(issuerDN))
{
    throw new PluginException($"Found multiple certificates issued by {issuerDN}");
}
```
to check the case. And added unit test.

Because we don't verify the chain, so it is possible to generate a fake DN loop. If the public key is verified, it is impossible to have a loop.